# Exercise 4.5: Data Consistency Checks

In [3]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [4]:
# Importing Orders Dataset
path = r'C:\Users\daxma\OneDrive\Desktop\Data Analytics\Data Immersion\Achievement 4\01-2025 Instacart Basket Analysis'
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [5]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [6]:
# Removing df_ords unnamed index column
df_ords = df_ords.drop(df_ords.columns[0], axis=1)

In [7]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [8]:
# Importing Product Dataset
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [9]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

### Mixed-Type Data

In [11]:
# Create a dataframe
df_test = pd.DataFrame()

In [12]:
# Create mixed type column
df_test['mix'] = ['a','b',1,True]

In [13]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [14]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].map(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


### Missing Values

In [16]:
# Finding missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [17]:
# Subset of missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [18]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [19]:
df_prods.shape

(49693, 5)

In [20]:
# Subset non-missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [21]:
df_prods_clean.shape

(49677, 5)

### Duplicates

In [23]:
# Finding full duplicates in df_prods_clean
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [24]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [25]:
df_prods_clean.shape

(49677, 5)

In [26]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [27]:
df_prods_clean_no_dups.shape

(49672, 5)

In [28]:
# Exporting cleaned prods data as csv
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'product_checked.csv'))

# Task 4.5

In [30]:
# Step 2 describe df_ords data and investigate results
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

#### Order Number: 
#### -Max: 100 orders. We can investigate if this is truly the max number of orders any single customer has placed or if the system stops counting after 100 orders.
#### Days since prior order:
#### -Max: 30 days. Are order gaps longer than 30 days being cut off? Some customers may order less frequently (e.g. every 45 days). Investigate if systems cuts off at 30 days.

In [32]:
# Step 3 Checking for mixed type data in df_ords
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].map(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [33]:
# Step 4 fixing mixed-type data

#### No output, therefore, there are no mixed-type columns in df_ords.

In [35]:
# Step 5 Check missing values in df_ords
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [36]:
# Subset of df_ords missing values
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [37]:
df_ords_nan

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
11        2168274        2             1                   2   
26        1374495        3             1                   1   
39        3343014        4             1                   6   
45        2717275        5             1                   3   
...           ...      ...           ...                 ...   
3420930    969311   206205             1                   4   
3420934   3189322   206206             1                   3   
3421002   2166133   206207             1                   6   
3421019   2227043   206208             1                   1   
3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 6 columns]

#### It appears that there are 206209 missing values in the days_since_prior_order column. Upon inspection it is clear that these missing values are due to the rows being the customers first order and therefore would not have any days since prior order.

In [39]:
# Step 6 Addressing missing values
# Create a new column 'first_order_flag' 
# If 'days_since_prior_order' is NaN, set flag to 1 (First Order), else 0 (Not First Order)
df_ords['first_order_flag'] = df_ords['days_since_prior_order'].isna().astype(int)

In [40]:
# Check if the column was created correctly
print(df_ords[['days_since_prior_order', 'first_order_flag']].head())

df_ords.head()

   days_since_prior_order  first_order_flag
0                     NaN                 1
1                    15.0                 0
2                    21.0                 0
3                    29.0                 0
4                    28.0                 0


order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  first_order_flag  
0                     NaN                 1  
1                    15.0                 0  
2                    21.0                 0  
3                    29.0                 0  
4                    28.0                 0

#### I used this method as it would not make sense to remove or impute these missing values as they are correct. If it is a customers first order there would be no days since prior order. The flag allows you to quickly see if it is the first order or not. That being said, is this redundant because we already have an order_number column?

In [42]:
# Step 7 Check for duplicate values in df_ords
df_ords_dups = df_ords[df_ords.duplicated()]

In [43]:
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, first_order_flag]
Index: []

In [44]:
df_ords_dups.shape

(0, 7)

In [45]:
# Step 8 Address duplicate rows

### There are no full duplicate rows in the df_ords dataframe and therefore do not need to be addressed.

In [47]:
#Step 9 Export the cleaned df_ords dataframe
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))